### DISCUSSION TITLE
Adding support for AWS-LC in the openssl plugin

### DISCUSSION THREAD

**Discussion by geedo0** [strongswan/strongswan#1878] [Ideas]
Hi all, I'd like to contribute changes to the openssl plugin so that it can be compiled against the [AWS-LC library](https://github.com/aws/aws-lc). Compiling against this library improves performance of cryptographic operations on x86-64 and aarch64 processors, provides customers a path for FIPS compliance, access to new cryptographic algorithms (e.g. post-quantum), and maintenence independent of OpenSSL.

AWS-LC is a fork of BoringSSL/OpenSSL and is largely API compatible with OpenSSL 1.1.1 but has a few missing/changed APIs that prevent successful compilation. There's already to make strongSwan compile against [BoringSSL](https://github.com/search?q=repo%3Astrongswan%2Fstrongswan%20BORINGSSL&type=code), similar changes are needed to support AWS-LC. Some of these API mismatches can be solved on the AWS-LC side by implementing them, but it would be good to get some input on which of these are necessary or are better left conditioned off in strongSwan. From my investigation, I found these symbols which compilation.

<table>
<tr>
<td>Symbol name</td>
<td>Purpose</td>
<td>Notes</td>
</tr>
<tr>
<td>BN_CTX_secure_new</td>
<td>"secure" means that memory gets zeroed out.</td>
<td>Removed in BoringSSL</td>
</tr>
<tr>
<td>BN_secure_new</td>
<td>''</td>
<td>Removed in BoringSSL</td>
</tr>
<tr>
<td>EVP_aes_128_ccm</td>
<td>Support for AES-CCM AEAD.</td>
<td>Removed in BoringSSL</td>
</tr>
<tr>
<td>EVP_aes_192_ccm</td>
<td>''</td>
<td>Removed in BoringSSL</td>
</tr>
<tr>
<td>EVP_aes_256_ccm</td>
<td>''</td>
<td>Removed in BoringSSL</td>
</tr>
<tr>
<td>EVP_chacha20_poly1305</td>
<td>Support for chacha20-poly1305 AEAD.</td>
<td>BoringSSL implements this using the EVP_AEAD interface instead of EVP_CIPHER.</td>
</tr>
<tr>
<td>EVP_shake128</td>
<td>Support for SHAKE in xof plugin.</td>
<td>Removed in BoringSSL</td>
</tr>
<tr>
<td>EVP_shake256</td>
<td>''</td>
<td>Removed in BoringSSL</td>
</tr>
<tr>
<td>RAND_priv_bytes</td>
<td>Alternative RAND_bytes method to call OpenSSL's private DRBG.</td>
<td>Removed in BoringSSL. AWS-LC has its own DRBG inside RAND_bytes.</td>
</tr>
<tr>
<td>SHA_CTX.h{0..4}</td>
<td>Hash data.</td>
<td>Changed to an array in BoringSSL</td>
</tr>
<tr>
<td>OPENSSL_KEY_FALLBACK</td>
<td>Fallback getter/setter symbols made opaque in OpenSSL 3</td>
<td>Macro logic needs to be fixed. AWS-LC is 1.1.1 based and does not need these fallback symbols.</td>
</tr>
<tr>
</table>

The other hurdle to overcome is how to deal with conflicts with system OpenSSL libraries. AWS-LC shares the same SONAME as OpenSSL. Typically, AWS-LC customers deal with this on a case by case basis by configuring their systems to dynamically link correctly and/or statically linking the binaries which they want to use with AWS-LC. I'm curious about any suggestions on what strategy is appropriate for strongSwan.

Would this work be okay for the strongSwan project?

**Comment by tobiasbrunner** [comment]
> Some of these API mismatches can be solved on the AWS-LC side by implementing them

Definitely preferable than additional `#ifdef`/`#if`s, but if they are relatively simple/short/contained, also fine.

> BN_CTX_secure_new, BN_secure_new

For older OpenSSL versions (< 1.1) we already use `BN_CTX_new` and `BN_new` as fallback (see the defines at the top of `src/libstrongswan/plugins/openssl/openssl_rsa_private_key.c`). Not sure, but I guess that already applies when compiling against BoringSSL as that announces an older version (at least I haven't seen any errors regarding that, as the change came in last year and we only switched from BoringSSL to OpenSSL for our Android app this year - although, we did use a rather old version of BoringSSL).

> EVP_aes_128_ccm, EVP_aes_192_ccm, EVP_aes_256_ccm

Similar to the above, these are used under an OpenSSL version condition (>= 1.1). So they might already not apply to BoringSSL.

> EVP_chacha20_poly1305
> BoringSSL implements this using the EVP_AEAD interface instead of EVP_CIPHER.

There is no EVP_AEAD in OpenSSL, so what interface are you referring to? `EVP_get_cipherbyname`? Or is that something specific to BoringSSL? If so, that might be the most significant API break.

The symbol is only used after a version (>= 1.1) and feature check (`!OPENSSL_NO_CHACHA`).

> EVP_shake128, EVP_shake256

These are also only used after a version check (>= 1.1.1) and `OPENSSL_NO_SHAKE` must not be defined.

Only the custom implementation of FrodoKEM uses these currently (well, in the master branch, the legacy _newhope_ plugin does too, but that will be removed soon). So that might not be much of an issue (OpenSSL currently does not provide a streaming API for these XOFs, so using them via that plugin is not ideal anyway).

> RAND_priv_bytes

Only used after a version check (>= 1.1.1).

> SHA_CTX.h{0..4}

Tricky anyway as OpenSSL 3 does not give access to this internal state unless compiled with deprecated APIs. Luckily, this is for a rarely used feature (EAP-AKA-3GPP2 is the only user of `PRF_KEYED_SHA1`).

> OPENSSL_KEY_FALLBACK
> Macro logic needs to be fixed. AWS-LC is 1.1.1 based and does not need these fallback symbols.

OK, that makes this awkward, but it explains why some of these symbols are an issue for your library but not for BoringSSL (at least not the version we used). What defines does BoringSSL or your library provide to distinguish this? (We already use `OPENSSL_IS_BORINGSSL` and `BORINGSSL_API_VERSION` in some places, are these usable for some of these?)

> Typically, AWS-LC customers deal with this on a case by case basis by configuring their systems to dynamically link correctly and/or statically linking the binaries which they want to use with AWS-LC. I'm curious about any suggestions on what strategy is appropriate for strongSwan.

I'd also say let the users deal with this (via C/LDFLAGS etc. when running configure, see e.g. [our test script](https://github.com/strongswan/strongswan/blob/master/scripts/test.sh#L130-L132)). Or what did you have in mind?



> Would this work be okay for the strongSwan project?

Sure, but, as always, depends on the details ;)

**Comment by geedo0** [comment]
> For older OpenSSL versions (< 1.1)  ... we did use a rather old version of BoringSSL).

This explains why a lot of the conditional logic is breaking for AWS-LC and not BoringSSL. AWS-LC defines this for the version number `OPENSSL_VERSION_NUMBER 0x1010107f` which by convention is OpenSSL 1.1.1 with `0x7f` for "patch" and "status" fields. The logic in strongSwan treats this as >1.1.1. Would it make sense to mask out these bits and only condition on the major/minor/fix numbers?

One thing I know about BoringSSL is that it initially targeted OpenSSL 1.0. Later versions upgraded BoringSSL to target OpenSSL 1.1.1 (Which is where AWS-LC is forked from). Do you have any context on the BoringSSL version integrated with strongSwan and some history on that effort?

> There is no EVP_AEAD in OpenSSL, so what interface are you referring to? EVP_get_cipherbyname? Or is that something specific to BoringSSL? If so, that might be the most significant API break.

You're right, this is purely a BoringSSL construct. They broke out the AEAD ciphers into their own interface [EVP_AEAD](https://github.com/aws/aws-lc/blob/fe308e3c76da834313f9f9c8f4114ef329cea034/crypto/fipsmodule/cipher/internal.h#L111) which has fields specific to AEAD. I'm not sure how different these are from EVP_CIPHER, but I agree that this is the most significant API break. Worst case, it will require implementing against the new interface. Best case, we can introduce a backwards compatible version that exposes an EVP_CIPHER version of the same algorithm.

> EVP_shake128, EVP_shake256 ... Only the custom implementation of FrodoKEM uses these currently (well, in the master branch, the legacy newhope plugin does too, but that will be removed soon). So that might not be much of an issue (OpenSSL currently does not provide a streaming API for these XOFs, so using them via that plugin is not ideal anyway).

Thanks for the useful context, these would've been difficult for us to implement on our end.

>  What defines does BoringSSL or your library provide to distinguish this? (We already use OPENSSL_IS_BORINGSSL and BORINGSSL_API_VERSION in some places, are these usable for some of these?)

BoringSSL's guidance says to use these defines to implement coniditonal logic. AWS-LC follows the same convention with `OPENSSL_IS_AWSLC` and `AWSLC_API_VERSION`.

> I'd also say let the users deal with this (via C/LDFLAGS etc. when running configure, see e.g. our test script). Or what did you have in mind?

This is exactly what I'd have in mind. Thanks for the test pointer!

> Sure, but, as always, depends on the details ;)

👍

**Reply by tobiasbrunner** [reply]
> Would it make sense to mask out these bits and only condition on the major/minor/fix numbers?

I don't think that would make a difference as we don't have any `>` checks, they are always `>=` or `<`.

> Do you have any context on the BoringSSL version integrated with strongSwan and some history on that effort?

Sure, you can find the old fork we used in the Anrdoid app in the _ndk-static_ branch at [strongswan/boringssl](https://github.com/strongswan/boringssl/tree/ndk-static). That branch uses `0x10002000` [as version number](https://github.com/strongswan/boringssl/blob/317b65a021794a614ad9e859910b83031fb1122a/src/include/openssl/base.h#L112).

A couple of years ago there was also some effort to make the _openssl_ plugin compatible with more recent versions (#352, you have to scroll down a bit for the actual discussion as it took some convincing to get me to look into it :wink:). But that was obviously before the work to make the _openssl_ plugin compatible with Openssl 3 last year. Again, the plugin still worked with our old version after that, but compatibility with newer versions might got broken. We currently don't have any CI (or manual) builds against recent versions of BoringSSL (it's officially still not recommended for public use, so there really was no reason to do so). Adding one against AWS-LC would probably be a good idea in this process.


> You're right, this is purely a BoringSSL construct. They broke out the AEAD ciphers into their own interface [EVP_AEAD](https://github.com/aws/aws-lc/blob/fe308e3c76da834313f9f9c8f4114ef329cea034/crypto/fipsmodule/cipher/internal.h#L111) which has fields specific to AEAD.

I see. Definitely looks like the cleaner interface.


> Worst case, it will require implementing against the new interface.

I don't think that's so bad. It's basically the member in the `private_aead_t` struct and `crypt()` that would be different (but I assume the latter would be shorter for this interface). For the constructor we can probably add some defines (e.g. `#define EVP_aes_128_gcm EVP_aead_aes_128_gcm` and exclude CCM via API checks). The implementation can also easily be tested as we have test vectors for all these algorithms.


> Thanks for the useful context, these would've been difficult for us to implement on our end.

Are there any plans to implement XOFs like SHAKE in BoringSSL or your library? (In OpenSSL there is still an open issue regarding possible interface changes for streaming output, see https://github.com/openssl/openssl/issues/7894).



> AWS-LC follows the same convention with `OPENSSL_IS_AWSLC` and `AWSLC_API_VERSION`.

I see. If these are roughly compatible to BoringSSL, it might then be easier to redefine the BoringSSL constants with these values if `OPENSSL_IS_AWSLC` is defined (maybe in `openssl_util.h`).

**Reply by geedo0** [reply]
Thanks a lot for the great feedback here, you saved me a lot of guesswork and bad decisions in the process.

> I don't think that would make a difference as we don't have any > checks, they are always >= or <.

I took a look at the source again and you're right. The incompatibility issues arise exclusively from mismatched interfaces.

> Adding one against AWS-LC would probably be a good idea in this process.

Agreed, we definitely want to prevent bitrot where possible.

> Are there any plans to implement XOFs like SHAKE in BoringSSL or your library?

Funny enough, I learned that a colleague is taking this on right now for something completely unrelated.

> I see. If these are roughly compatible to BoringSSL, it might then be easier to redefine the BoringSSL constants with these values if OPENSSL_IS_AWSLC is defined (maybe in openssl_util.h).

Makes sense to me.

**Comment by geedo0** [comment]
Closing this thread in favor of the corresponding [issue](https://github.com/strongswan/strongswan/issues/1907).
